diff --git a/clang/lib/Driver/ToolChains/Gnu.cpp b/clang/lib/Driver/ToolChains/Gnu.cpp
--- a/clang/lib/Driver/ToolChains/Gnu.cpp
+++ b/clang/lib/Driver/ToolChains/Gnu.cpp
@@ -2396,6 +2396,14 @@ void tools::gnuclang::Linker::ConstructJob(Compilation &C, const JobAction &JA,
     else
       CmdArgs.push_back("--no-relax");
   }
+
+  // llvmbox: Always use --as-needed for musl static linking
+  CmdArgs.push_back("--as-needed");
+
+  // llvmbox: Use lld as in-process linker when targeting musl
+  if (getToolChain().getTriple().isMusl()) {
+    CmdArgs.push_back("-fuse-ld=lld");
+  }
 }
 #endif
